File: source/test/java/com/xeomar/annex/ProgramTest.java
package com.xeomar.annex;

import com.xeomar.product.ProductCard;
import com.xeomar.util.LogFlag;
import com.xeomar.util.LogUtil;
import com.xeomar.util.Parameters;
import org.junit.Test;

import java.io.*;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.assertThat;

public class ProgramTest {

	private static final String CHARSET = "utf-8";

	@Test
	public void testProductCard() {
		ProductCard card = new Program().getCard();
		assertThat( card, not( is( nullValue() ) ) );
	}

	@Test( timeout = 2000 )
	public void testReadCommandsFromStdin() throws IOException {
		InputStream originalInput = System.in;
		PrintStream originalOutput = System.out;

		PipedInputStream outputPipe = new PipedInputStream();
		PipedOutputStream inputPipe = new PipedOutputStream();

		System.setIn( new PipedInputStream( inputPipe ) );
		System.setOut( new PrintStream( new PipedOutputStream( outputPipe ) ) );
		try {
			new Thread( () -> {
				try {
					new Program().start( new String[]{ UpdateFlag.STDIN, LogFlag.LOG_LEVEL, "none" } );
				} catch( Exception exception ) {
					exception.printStackTrace( System.err );
				}
			} ).start();

			inputPipe.write( "pause 13\n".getBytes( CHARSET ) );
			inputPipe.write( "pause 7\n".getBytes( CHARSET ) );
			inputPipe.close();
			String result1 = new BufferedReader( new InputStreamReader( outputPipe, CHARSET ) ).readLine();
			assertThat( result1, is( "SUCCESS paused 13ms" ) );
			String result2 = new BufferedReader( new InputStreamReader( outputPipe, CHARSET ) ).readLine();
			assertThat( result2, is( "SUCCESS paused 7ms" ) );
		} finally {
			// Restore the original streams
			System.setOut( originalOutput );
			System.setIn( originalInput );
		}
	}

	@Test( timeout = 2000 )
	public void testReadCommandsFromBytes() throws IOException {
		PipedReader outputPipe = new PipedReader();
		PipedWriter inputPipe = new PipedWriter();

		PipedReader reader = new PipedReader( inputPipe );
		PipedWriter writer = new PipedWriter( outputPipe );

		new Thread( () -> {
			try {
				Program program = new Program();
				LogUtil.configureLogging( program, Parameters.parse( LogFlag.LOG_LEVEL, "none" ) );
				program.runTasksFromReader( reader, writer );
			} catch( Exception exception ) {
				exception.printStackTrace( System.err );
			}
		} ).start();

		inputPipe.write( "pause 17\n" );
		inputPipe.write( "pause 5" );
		inputPipe.close();

		String result1 = new BufferedReader( outputPipe ).readLine();
		assertThat( result1, is( "SUCCESS paused 17ms" ) );

		String result2 = new BufferedReader( outputPipe ).readLine();
		assertThat( result2, is( "SUCCESS paused 5ms" ) );
	}

}

File: source/test/java/com/xeomar/annex/ProgramTest.kt
package com.xeomar.annex

import kotlinx.coroutines.experimental.launch
import org.hamcrest.Matchers.`is`
import org.junit.Assert.assertThat
import org.junit.Test
import java.io.*
import java.nio.charset.Charset

class ProgramTest {

	@Test
	@Throws(Exception::class)
	fun testReadCommandsFromStdin() {
		val originalInput = System.`in`
		val originalOutput = System.out

		val outputPipe = PipedInputStream()
		val inputPipe = PipedOutputStream()

		System.setIn(PipedInputStream(inputPipe))
		System.setOut(PrintStream(PipedOutputStream(outputPipe)))
		try {
			launch {
				Program.run(arrayOf("--stream"))
			}

			inputPipe.write("update source target\n".toByteArray(charset("utf-8")))
			inputPipe.flush()
			val result1 = TaskResult.parse(outputPipe.bufferedReader(charset("utf-8")).readLine())
			assertThat(result1.code, `is`(TaskStatus.SUCCESS))
			assertThat(result1.message, `is`("success"))

			inputPipe.write("update source2 target2".toByteArray(charset("utf-8")))
			inputPipe.close()
			val result2 = TaskResult.parse(outputPipe.bufferedReader(charset("utf-8")).readLine())
			assertThat(result2.code, `is`(TaskStatus.SUCCESS))
			assertThat(result2.message, `is`("success"))
		} finally {
			// Restore the original streams
			System.setOut(originalOutput)
			System.setIn(originalInput)
		}
	}

	@Test
	@Throws(Exception::class)
	fun testReadCommandsFromBytes() {
		val outputPipe = PipedReader()
		val inputPipe = PipedWriter()

		val reader = PipedReader(inputPipe)
		val writer = PipedWriter(outputPipe)

		launch {
			Program.runTasksFromReader(reader, writer)
		}

		inputPipe.write("update source target\n")
		inputPipe.flush()
		val result1 = TaskResult.parse(outputPipe.buffered().readLine())
		assertThat(result1.code, `is`(TaskStatus.SUCCESS))
		assertThat(result1.message, `is`("success"))

		inputPipe.write("update source2 target2")
		inputPipe.close()
		val result2 = TaskResult.parse(outputPipe.buffered().readLine())
		assertThat(result2.code, `is`(TaskStatus.SUCCESS))
		assertThat(result2.message, `is`("success"))
	}

}
